torch/csrc/jit/codegen/cuda/expr_evaluator.h

```diff
@@ -72,11 +74,67 @@ class TORCH_CUDA_API ExpressionEvaluator : private IterVisitor {
   void handle(kir::UnaryOp*) override;
   void handle(kir::BinaryOp*) override;
 
- private:
+ protected:
   const EvaluationContext* context_ = nullptr;
   std::unordered_map<const Statement*, Int::ScalarType> values_;
 };
 
+class TORCH_CUDA_API StatefulExpressionEvaluator : private OptOutDispatch {
+ public:
+  explicit StatefulExpressionEvaluator(Fusion* fusion) : fusion_(fusion) {}
+
+  Fusion* fusion() const {
+    return fusion_;
+  }
+
+  void safeBind(Val* value, Int::ScalarType concrete_value);
+
+  // Returns value if found in mapping, otherwise returns c10::nullopt
+  c10::optional<Int::ScalarType> getValue(Val* value);
+
+  // Checks if value is already infered, returns infered value if so, otherwise
+  // runs traversal on value. Warning: should not be called in traversal.
+  c10::optional<Int::ScalarType> inferValue(Val* value);
+
+ private:
+  std::unordered_map<const Val*, Int::ScalarType> bindings_;
+  Fusion* fusion_ = nullptr;
+
+  using OptOutDispatch::handle;
+
+  void handle(Expr* expr) override {
```

Comment:
please place the private methods in a dedicated private: section (so the private state is always last in a class definition)

Reply:
Blah, fine.